File: tnzhou/253.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        unordered_map <int, int> m;
        vector<int> timePoints;
        for (int i = 0; i < intervals.size(); i++) {
            timePoints.push_back(intervals[i].start);
            timePoints.push_back(intervals[i].end);
            m[intervals[i].start]++;
            m[intervals[i].end]--;
        }
        sort(timePoints.begin(), timePoints.end());
        int maxCnt = 0;
        int cnt = 0;
        for (int i = 0; i < timePoints.size(); i++) {
            if (i == 0 || timePoints[i] != timePoints[i - 1]) {
                cnt += m[timePoints[i]];
                if (cnt > maxCnt)
                    maxCnt = cnt;
            }
        }
                                
        return maxCnt;                        
    }
};

File: zemiao/253.py
"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""
import heapq
class Solution:
    """
    @param intervals: an array of meeting time intervals
    @return: the minimum number of conference rooms required
    """
    def minMeetingRooms(self, intervals):
        # Write your code here
        heap = []
        intervals.sort(key=lambda x:x.start)
        for x in intervals:
          if heap and heap[0][1].end <= x.start:
            heap[0] = (x.end, x)
            heapq.heapify(heap)
            continue
          
          heapq.heappush(heap,(x.end, x))
        return len(heap)